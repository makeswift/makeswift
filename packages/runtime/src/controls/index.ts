export * from './control'

export {
  ControlDefinition,
  Checkbox,
  CheckboxDefinition,
  Color,
  ColorDefinition,
  Combobox,
  ComboboxDefinition,
  Font,
  FontDefinition,
  Group,
  GroupDefinition,
  GroupControl,
  IconRadioGroupDefinition,
  type IconRadioGroupIcon,
  unstable_IconRadioGroup,
  Image,
  ImageDefinition,
  List,
  ListDefinition,
  ListControl,
  Number,
  NumberDefinition,
  Select,
  SelectDefinition,
  Shape,
  ShapeDefinition,
  ShapeControl,
  Style,
  StyleDefinition,
  type StyleProperty,
  StyleControl,
  TextInput,
  TextInputDefinition,
  TextArea,
  TextAreaDefinition,
  unstable_Typography,
  unstable_TypographyDefinition,
} from '@makeswift/controls'

export { Link, LinkDefinition } from './link'
export { RichTextV1Definition, RichTextV1Control } from './rich-text'
export { RichText, RichTextV2Definition, RichTextV2Control } from './rich-text-v2'
export { Slot, SlotDefinition, SlotControl } from './slot'
export { unstable_StyleV2, StyleV2Definition, StyleV2Control } from './style-v2/style-v2'
