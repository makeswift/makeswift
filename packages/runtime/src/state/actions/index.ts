import { isAction } from '@reduxjs/toolkit'

import { type BuilderAction, BuilderActionTypes } from '../builder-api'
import { type HostAction, HostActionTypes } from '../host-api'

import { type InternalAction, InternalActionTypes } from './internal'

export { type UnknownAction } from '@reduxjs/toolkit'
export { type DocumentPayload } from '../shared-api'

export const ActionTypes = {
  ...HostActionTypes,
  ...BuilderActionTypes,
  ...InternalActionTypes,
} as const

export type Action = HostAction | BuilderAction | InternalAction

export function isKnownAction(action: unknown): action is Action {
  return isAction(action) && Object.hasOwn(ActionTypes, action.type)
}
