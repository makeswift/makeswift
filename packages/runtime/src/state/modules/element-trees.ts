import { type Operation } from 'ot-json0'
import { getIn } from 'immutable'

import { type Element, type ElementData, isElementReference } from '@makeswift/controls'

import { Introspection } from '../../prop-controllers'

import { type Action, ActionTypes } from '../actions'

import { getRootElement, type Document } from './read-only-documents'
import { type DescriptorsByComponentType } from './prop-controllers'

export type ElementTree = {
  elements: Map<string, Element>
  elementIds: Map<string, string>
}

export type State = Map<string, ElementTree>

export function getInitialState(
  documents?: Map<string, Document>,
  descriptors?: DescriptorsByComponentType,
): State {
  const state = new Map<string, ElementTree>()
  if (documents == null || descriptors == null) return state

  for (const [documentKey, document] of documents) {
    state.set(documentKey, buildElementTree(getRootElement(document), descriptors))
  }

  return state
}

function getElementTree(state: State, id: string): ElementTree | null {
  return state.get(id) ?? null
}

export function getElements(state: State, documentKey: string): Map<string, Element> {
  return getElementTree(state, documentKey)?.elements ?? new Map()
}

export function getElementIds(state: State, documentKey: string): Map<string, string> {
  return getElementTree(state, documentKey)?.elementIds ?? new Map()
}

export function getElement(state: State, documentKey: string, elementKey: string): Element | null {
  return getElements(state, documentKey).get(elementKey) ?? null
}

export function getElementId(state: State, documentKey: string, elementKey: string): string | null {
  return getElementIds(state, documentKey).get(elementKey) ?? null
}

export function reducer(state: State = getInitialState(), action: Action): State {
  switch (action.type) {
    case ActionTypes.CREATE_ELEMENT_TREE: {
      const { document, descriptors } = action.payload
      return new Map(state).set(
        document.key,
        buildElementTree(getRootElement(document), descriptors),
      )
    }

    case ActionTypes.DELETE_ELEMENT_TREE: {
      const nextState = new Map(state)
      const deleted = nextState.delete(action.payload.documentKey)
      return deleted ? nextState : state
    }

    case ActionTypes.CHANGE_ELEMENT_TREE: {
      const { oldDocument, newDocument, descriptors, operation } = action.payload
      const documentKey = oldDocument.key
      console.assert(
        documentKey === newDocument.key,
        `Mismatching document keys ${documentKey} !== ${newDocument.key}`,
      )

      const elementTree = state.get(documentKey)
      if (elementTree == null) return state

      const updatedElementTree = applyChanges(
        elementTree,
        descriptors,
        {
          old: getRootElement(oldDocument),
          new: getRootElement(newDocument),
        },
        operation,
      )

      return new Map(state).set(documentKey, updatedElementTree)
    }

    default:
      return state
  }
}

export function* traverseElementTree(
  element: Element,
  descriptors: DescriptorsByComponentType,
): Generator<Element> {
  yield element
  if (isElementReference(element)) return

  const elementDescriptors = descriptors.get(element.type)
  if (elementDescriptors == null) return

  for (const [propKey, descriptor] of Object.entries(elementDescriptors)) {
    const children = Introspection.getElementChildren(descriptor, element.props[propKey])
    for (const child of children) {
      yield* traverseElementTree(child, descriptors)
    }
  }
}

function getElementIdProp(
  element: ElementData,
  descriptors: DescriptorsByComponentType,
): string | null {
  const elementDescriptors = descriptors.get(element.type)
  if (elementDescriptors == null) return null

  for (const [propName, descriptor] of Object.entries(elementDescriptors)) {
    const elementId = Introspection.getElementId(descriptor, element.props[propName])
    if (elementId != null) return elementId
  }

  return null
}

export function buildElementTree(
  rootElement: Element,
  descriptors: DescriptorsByComponentType,
): ElementTree {
  const elements = new Map<string, Element>()
  const elementIds = new Map<string, string>()

  for (const element of traverseElementTree(rootElement, descriptors)) {
    elements.set(element.key, element)
    if (!isElementReference(element)) {
      const elementId = getElementIdProp(element, descriptors)
      if (elementId != null) elementIds.set(element.key, elementId)
    }
  }

  return {
    elements,
    elementIds,
  }
}

type OperationPath = Operation[number]['p']

// performance-sensitive function, intentionally not using `elementData` schema here
function isElement(item: unknown): item is Element {
  return (
    typeof item === 'object' &&
    item != null &&
    'key' in item &&
    'type' in item &&
    typeof item.key === 'string' &&
    typeof item.type === 'string'
  )
}

type ElementOperationPath = { elementPath: OperationPath; propName: string }

function getElementOperationPath(path: OperationPath): ElementOperationPath {
  const i = path.findLastIndex(
    (fragment, i) =>
      typeof fragment === 'number' && (i === path.length - 1 || path[i + 1] === 'props'),
  )

  if (path.length === 0 || (i === -1 && path[0] !== 'props') || path.length - i < 3) {
    console.error('Operation path does not point to an element property', { path })
    return { elementPath: [], propName: 'children' }
  }

  return { elementPath: path.slice(0, i + 1), propName: `${path.slice(i + 1).at(1)}` }
}

export function getChangedElementsPaths(
  path: OperationPath,
): [ElementOperationPath, ...ElementOperationPath[]] {
  let elementOp = getElementOperationPath(path)
  const result: [ElementOperationPath, ...ElementOperationPath[]] = [elementOp]
  while (elementOp.elementPath.length > 0) {
    elementOp = getElementOperationPath(elementOp.elementPath.slice(0, -1))
    result.push(elementOp)
  }

  return result
}

function getElementByPath(rootElement: Element, elementPath: OperationPath): Element | null {
  const item = getIn(rootElement, elementPath)
  if (!isElement(item)) {
    console.error('Expected an element, got', item, {
      rootElement,
      elementPath,
    })

    return null
  }

  return item
}

function getElementAndPropName(
  rootElement: Element,
  { elementPath, propName }: ElementOperationPath,
): [Element, string] {
  const element = getElementByPath(rootElement, elementPath)
  return element != null ? [element, propName] : [rootElement, 'children']
}

function updateParentElements(
  elements: ElementTree['elements'],
  elementPaths: ElementOperationPath[],
  rootElement: Element,
): void {
  elementPaths.forEach(({ elementPath }) => {
    const element = getElementByPath(rootElement, elementPath)
    if (element != null) elements.set(element.key, element)
  })
}

function hasChildren(
  element: Element,
  propName: string,
  descriptors: DescriptorsByComponentType,
): boolean {
  if (isElementReference(element)) return false

  const propDescriptor = descriptors.get(element.type)?.[propName]
  if (propDescriptor == null) return false

  const children = Introspection.getElementChildren(propDescriptor, element.props[propName])
  return children.length > 0
}

function deleteElement(
  { elements, elementIds }: ElementTree,
  deletedElement: Element,
  propName: string,
  descriptors: DescriptorsByComponentType,
) {
  if (hasChildren(deletedElement, propName, descriptors)) {
    for (const element of traverseElementTree(deletedElement, descriptors)) {
      elements.delete(element.key)
      elementIds.delete(element.key)
    }
  } else {
    elements.delete(deletedElement.key)
    elementIds.delete(deletedElement.key)
  }
}

function applyDelete(
  elementTree: ElementTree,
  descriptors: DescriptorsByComponentType,
  rootElements: { old: Element; new: Element },
  path: OperationPath,
): ElementTree {
  const [deleteElementPath, ...parentElementPaths] = getChangedElementsPaths(path)
  const [deletedElement, propName] = getElementAndPropName(rootElements.old, deleteElementPath)

  const elements = new Map(elementTree.elements)
  const elementIds = new Map(elementTree.elementIds)

  deleteElement({ elements, elementIds }, deletedElement, propName, descriptors)
  updateParentElements(elements, parentElementPaths, rootElements.new)

  return {
    elements,
    elementIds,
  }
}

function insertElement(
  { elements, elementIds }: ElementTree,
  insertedElement: Element,
  propName: string,
  descriptors: DescriptorsByComponentType,
) {
  if (hasChildren(insertedElement, propName, descriptors)) {
    for (const element of traverseElementTree(insertedElement, descriptors)) {
      elements.set(element.key, element)
      if (!isElementReference(element)) {
        const elementId = getElementIdProp(element, descriptors)
        if (elementId != null) elementIds.set(element.key, elementId)
      }
    }
  } else {
    elements.set(insertedElement.key, insertedElement)
    if (!isElementReference(insertedElement)) {
      const elementId = getElementIdProp(insertedElement, descriptors)
      if (elementId != null) elementIds.set(insertedElement.key, elementId)
    }
  }
}

function applyInsert(
  elementTree: ElementTree,
  descriptors: DescriptorsByComponentType,
  rootElements: { old: Element; new: Element },
  path: OperationPath,
): ElementTree {
  const [insertedElementPath, ...parentElementPaths] = getChangedElementsPaths(path)
  const [insertedElement, propName] = getElementAndPropName(rootElements.new, insertedElementPath)

  const elements = new Map(elementTree.elements)
  const elementIds = new Map(elementTree.elementIds)

  insertElement({ elements, elementIds }, insertedElement, propName, descriptors)
  updateParentElements(elements, parentElementPaths, rootElements.new)

  return {
    elements,
    elementIds,
  }
}

function applyUpdate(
  elementTree: ElementTree,
  descriptors: DescriptorsByComponentType,
  rootElements: { old: Element; new: Element },
  path: OperationPath,
): ElementTree {
  const [updateElementPath, ...parentElementPaths] = getChangedElementsPaths(path)
  const [deletedElement, propName] = getElementAndPropName(rootElements.old, updateElementPath)
  const [insertedElement, _] = getElementAndPropName(rootElements.new, updateElementPath)

  const elements = new Map(elementTree.elements)
  const elementIds = new Map(elementTree.elementIds)

  deleteElement({ elements, elementIds }, deletedElement, propName, descriptors)
  insertElement({ elements, elementIds }, insertedElement, propName, descriptors)

  updateParentElements(elements, parentElementPaths, rootElements.new)

  return {
    elements,
    elementIds,
  }
}

function applyChanges(
  elementTree: ElementTree,
  descriptors: DescriptorsByComponentType,
  rootElements: { old: Element; new: Element },
  operation: Operation,
): ElementTree {
  return operation.reduce((tree, op) => {
    const hasDelete = 'ld' in op || 'od' in op
    const hasInsert = 'li' in op || 'oi' in op
    if (hasDelete && hasInsert) {
      return applyUpdate(tree, descriptors, rootElements, op.p)
    }

    if (hasDelete) return applyDelete(tree, descriptors, rootElements, op.p)
    if (hasInsert) return applyInsert(tree, descriptors, rootElements, op.p)
    return tree
  }, elementTree)
}
