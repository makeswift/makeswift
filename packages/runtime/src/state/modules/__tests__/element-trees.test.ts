import * as ElementTrees from '../element-trees'
import * as Fixtures from './fixtures/element-trees'
import {
  ELEMENT_TREE_DEMO_COMPONENT_TYPE,
  ElementTreesDemo,
} from './fixtures/element-trees-demo-component'

import { getPropControllerDescriptors } from '../../react-page'
import { ReactRuntime } from '../../../runtimes/react/react-runtime'
import { Slot } from '../../../controls'
import * as Actions from '../../actions'

describe('traverseElementTree', () => {
  const runtime = new ReactRuntime()

  test('walks the tree in the correct order', () => {
    const descriptors = getPropControllerDescriptors(runtime.store.getState())
    const elements = ElementTrees.traverseElementTree(Fixtures.homePage, descriptors)
    expect([...elements].map(({ key, type }) => ({ key, type }))).toMatchSnapshot()
  })
})

describe('getChangedElementsPaths', () => {
  test('gracefully handles empty path', () => {
    const changedPaths = ElementTrees.getChangedElementsPaths([])
    expect(changedPaths).toStrictEqual([{ elementPath: [], propName: null }])
  })

  describe('gracefully handles incorrect paths', () => {
    let consoleError: jest.SpyInstance
    beforeEach(() => {
      consoleError = jest.spyOn(console, 'error').mockImplementation(jest.fn())
    })

    afterEach(() => {
      expect(console.error).toHaveBeenCalled()
      consoleError.mockRestore()
    })

    test.each([
      { path: [0] },
      { path: ['foo', 0] },
      { path: ['foo', 0, 'props'] },
      { path: ['props', 'background', 7] },
    ])('$path', ({ path }) => {
      expect(ElementTrees.getChangedElementsPaths(path)).toStrictEqual([
        { elementPath: [], propName: null },
      ])
    })

    test("['foo', 0, 'props', 'background']", () => {
      expect(ElementTrees.getChangedElementsPaths(['foo', 0, 'props', 'background'])).toStrictEqual(
        [
          { elementPath: ['foo', 0], propName: 'background' },
          { elementPath: [], propName: null },
        ],
      )
    })
  })

  test('correctly partitions operation path', () => {
    expect(ElementTrees.getChangedElementsPaths(['props', 'margin'])).toStrictEqual([
      { elementPath: [], propName: 'margin' },
    ])

    expect(ElementTrees.getChangedElementsPaths(['props', 'margin', 0, 'foo'])).toStrictEqual([
      { elementPath: [], propName: 'margin' },
    ])

    expect(
      ElementTrees.getChangedElementsPaths(['props', 'children', 'value', 'elements', 4, 'foo']),
    ).toStrictEqual([{ elementPath: [], propName: 'children' }])

    expect(
      ElementTrees.getChangedElementsPaths([
        'props',
        'columns',
        'value',
        'elements',
        4,
        'props',
        'foo',
      ]),
    ).toStrictEqual([
      { elementPath: ['props', 'columns', 'value', 'elements', 4], propName: 'foo' },
      { elementPath: [], propName: 'columns' },
    ])

    expect(
      ElementTrees.getChangedElementsPaths([
        'props',
        'children',
        'value',
        'elements',
        1,
        'props',
        'children',
        'value',
        'elements',
        0,
        'props',
        'children',
      ]),
    ).toStrictEqual([
      {
        elementPath: [
          'props',
          'children',
          'value',
          'elements',
          1,
          'props',
          'children',
          'value',
          'elements',
          0,
        ],
        propName: 'children',
      },
      { elementPath: ['props', 'children', 'value', 'elements', 1], propName: 'children' },
      { elementPath: [], propName: 'children' },
    ])
  })
})

describe('resetting an element tree', () => {
  const runtime = new ReactRuntime()
  runtime.registerComponent(ElementTreesDemo, {
    type: ELEMENT_TREE_DEMO_COMPONENT_TYPE,
    label: 'Demo',
    props: {
      left: Slot(),
      right: Slot(),
    },
  })

  test('garbage collects all elements that were removed on reset', () => {
    const descriptors = getPropControllerDescriptors(runtime.store.getState())
    const documentKey = '0f567942-3ef5-4e66-abb0-969044145606'

    const oldDocument = {
      key: documentKey,
      rootElement: Fixtures.resetElementTree.beforeReset,
    }

    const initialState = ElementTrees.reducer(
      ElementTrees.getInitialState(),
      Actions.createElementTree({
        document: oldDocument,
        descriptors,
      }),
    )

    const initialElementKeys = Array.from(
      ElementTrees.getElements(initialState, documentKey).keys(),
    )
    expect(initialElementKeys).toMatchSnapshot()

    const newDocument = {
      key: documentKey,
      rootElement: Fixtures.resetElementTree.afterReset,
    }

    const resetOp = [{ p: [], od: oldDocument.rootElement, oi: newDocument.rootElement }]

    const updatedState = ElementTrees.reducer(
      initialState,
      Actions.changeElementTree({
        oldDocument,
        newDocument,
        descriptors,
        operation: resetOp,
      }),
    )

    const elementKeysAfterReset = Array.from(
      ElementTrees.getElements(updatedState, documentKey).keys(),
    )
    expect(elementKeysAfterReset).toMatchSnapshot()

    // The before/after fixtures are composed of entirely unique elements, so
    // they should share no common keys
    expect(initialElementKeys.some(key => elementKeysAfterReset.includes(key))).toBe(false)
    expect(elementKeysAfterReset.some(key => initialElementKeys.includes(key))).toBe(false)
  })
})
