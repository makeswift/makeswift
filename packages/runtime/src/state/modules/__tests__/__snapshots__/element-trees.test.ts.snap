// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`resetting an element tree garbage collects all elements that were removed on reset 1`] = `
[
  "e4e70b30-af46-5f25-93ee-2c550ec4a5f4",
  "51115f2a-12ac-4a0f-8be3-7e4006cf7c09",
  "f46bb567-8964-4606-b544-295b4b7d05fd",
  "c5010c7f-2eec-4042-abc8-be48836f8aef",
  "7039524a-c2a1-42be-9988-a799c21fe734",
  "618f069e-9515-4546-8abc-75cec6d33af4",
  "9745df8d-1589-46fd-8320-f3238bb94cbe",
  "92fd0115-ff8f-4218-b23f-9f861166ef35",
]
`;

exports[`resetting an element tree garbage collects all elements that were removed on reset 2`] = `
[
  "38bf059f-9a17-4626-bec4-e286130fd9da",
  "6ddc735a-3098-410a-b3b6-801f4ae1a0be",
]
`;

exports[`traverseElementTree walks the tree in the correct order 1`] = `
[
  {
    "key": "1cbb8e34-bf1c-4449-a002-71177d2f61ab",
    "type": "./components/Root/index.js",
  },
  {
    "key": "182a5df5-2ce8-4c26-8c6a-181aaa071e9c",
    "type": "./components/Box/index.js",
  },
  {
    "key": "1556e2db-f134-49b6-8091-f9b972c84a8e",
    "type": "./components/Navigation/index.js",
  },
  {
    "key": "f83c4c81-d23e-4ecf-857f-4dadc5e5fbd1",
    "type": "./components/Box/index.js",
  },
  {
    "key": "2de2123b-bf6c-4049-8d32-747d5451ccd0",
    "type": "./components/Box/index.js",
  },
  {
    "key": "fb01353b-ef6c-4fe3-a72f-6b05556f79ab",
    "type": "./components/Text/index.js",
  },
  {
    "key": "acf27ae9-21df-4701-8efc-ac1c2900f3ac",
    "type": "./components/Text/index.js",
  },
  {
    "key": "45c4bd3a-c398-4989-93b4-58900dbe4490",
    "type": "./components/Button/index.js",
  },
  {
    "key": "c1eaa434-28a6-4791-843c-c44d6672a8b6",
    "type": "./components/Box/index.js",
  },
  {
    "key": "ffe7d557-7d12-495a-9e80-dc2565e57f04",
    "type": "./components/Box/index.js",
  },
  {
    "key": "a2e9807f-f684-40c1-98f1-946676946cc2",
    "type": "./components/Box/index.js",
  },
  {
    "key": "e4b17748-47bf-479e-8230-e3f3eb9184b6",
    "type": "./components/Image/index.js",
  },
  {
    "key": "7b83e225-d6d4-4587-8776-42929c4b7992",
    "type": "./components/Text/index.js",
  },
  {
    "key": "d81b4dd7-8f45-4524-aa93-d58f6c7b322f",
    "type": "./components/Text/index.js",
  },
  {
    "key": "288b5f11-70ea-48d4-80ca-b8a690685f72",
    "type": "./components/Box/index.js",
  },
  {
    "key": "78458aea-6606-4395-9eb9-7720518c9282",
    "type": "./components/Image/index.js",
  },
  {
    "key": "5489494f-6c21-44fa-a67a-1898c9a75c27",
    "type": "./components/Text/index.js",
  },
  {
    "key": "7eb43920-04a6-4be9-ba81-2d08a4562feb",
    "type": "./components/Text/index.js",
  },
  {
    "key": "6eda79f6-8ba2-43a9-9973-bf433b92f29b",
    "type": "./components/Box/index.js",
  },
  {
    "key": "61276a5f-941c-4aef-908c-29593a48c68a",
    "type": "./components/Image/index.js",
  },
  {
    "key": "8e6a96dd-c598-4957-8e5d-4d96677d7797",
    "type": "./components/Text/index.js",
  },
  {
    "key": "465716fa-4256-4a4b-9edb-49d70f5a8b47",
    "type": "./components/Text/index.js",
  },
  {
    "key": "8b2248af-2de6-4d9d-84c1-d725701bec3d",
    "type": "./components/Box/index.js",
  },
  {
    "key": "a6add9b7-16ba-4d2b-b7ad-42512824a3aa",
    "type": "./components/Box/index.js",
  },
  {
    "key": "50070ac1-1832-4610-9e6e-9e64bcc302d8",
    "type": "./components/Box/index.js",
  },
  {
    "key": "ba5f8e72-9521-4d56-98ef-2416bd9de82f",
    "type": "./components/Image/index.js",
  },
  {
    "key": "ee1bf334-96cb-4fef-8178-4223f21e4c98",
    "type": "./components/Box/index.js",
  },
  {
    "key": "f4d9f509-aa1d-40c4-8685-c7dfb9af7f21",
    "type": "./components/Text/index.js",
  },
  {
    "key": "bdaeccb8-2ea1-404f-80ba-e366632d339a",
    "type": "./components/Text/index.js",
  },
  {
    "key": "8f44a1ea-8c4a-431f-a5ac-09d93ce189de",
    "type": "./components/Button/index.js",
  },
  {
    "key": "37506080-292b-4534-a21c-6ffcc10cd751",
    "type": "./components/Box/index.js",
  },
  {
    "key": "fb50cf71-6765-424a-9b68-00bec94d9262",
    "type": "./components/Box/index.js",
  },
  {
    "key": "3bbad17b-3ab5-4893-acdb-426d6a731c84",
    "type": "./components/Text/index.js",
  },
  {
    "key": "0c45f6fa-4a08-40a4-b2ac-c460d64d95ff",
    "type": "./components/Text/index.js",
  },
  {
    "key": "a554a115-b3b0-4e6a-b211-292a6ad1c6ed",
    "type": "./components/Button/index.js",
  },
  {
    "key": "687fb75a-fa88-437d-b2af-c08a639dabc7",
    "type": "./components/Image/index.js",
  },
  {
    "key": "154f4c19-941b-45f5-ba01-44a90aebe84f",
    "type": "./components/Box/index.js",
  },
  {
    "key": "29e84b30-417a-46ed-a852-68027e0c1de4",
    "type": "./components/Box/index.js",
  },
  {
    "key": "f77fe602-1619-4f21-8f44-6c93e5bbcf01",
    "type": "./components/Text/index.js",
  },
  {
    "key": "844ecfd2-e36b-417c-b9c8-f449573d7726",
    "type": "./components/Box/index.js",
  },
  {
    "key": "095033fc-0812-4bde-aa5c-aa178bec13d5",
    "type": "./components/Image/index.js",
  },
  {
    "key": "38cdf329-3b63-4b70-8d8c-186e1233df75",
    "type": "./components/Text/index.js",
  },
  {
    "key": "b7e673c7-41e3-40d2-8247-61f82acc6a82",
    "type": "./components/Text/index.js",
  },
  {
    "key": "e134a9e2-1300-4312-b2ba-119c91b6bf24",
    "type": "./components/Text/index.js",
  },
  {
    "key": "c6947e5b-af79-4d0f-9473-9e32db062f49",
    "type": "./components/Button/index.js",
  },
  {
    "key": "49687195-b717-407c-a820-05b83a37a06c",
    "type": "./components/Box/index.js",
  },
  {
    "key": "34901d6f-3df0-4cac-a8c9-150ff7296a95",
    "type": "./components/Image/index.js",
  },
  {
    "key": "7a096a31-93eb-42d2-9d72-1ce3eba628c6",
    "type": "./components/Text/index.js",
  },
  {
    "key": "fe5dd958-f9c4-44da-869b-bd6e86da7c8d",
    "type": "./components/Text/index.js",
  },
  {
    "key": "74447549-e0c7-4bb0-96a6-550571bcd378",
    "type": "./components/Text/index.js",
  },
  {
    "key": "debfe56a-e8a2-432e-aa95-edc316be18a4",
    "type": "./components/Button/index.js",
  },
  {
    "key": "d7b40aa9-41f0-4cc1-9cd8-6dfb86cbc849",
    "type": "./components/Box/index.js",
  },
  {
    "key": "6e1df1ae-7327-4a5f-b4c7-ae5b27d73bb4",
    "type": "./components/Image/index.js",
  },
  {
    "key": "bd194792-9843-42eb-9c9e-45962fdce3eb",
    "type": "./components/Text/index.js",
  },
  {
    "key": "067066c3-d9fe-4fc0-9e84-f017f3df2366",
    "type": "./components/Text/index.js",
  },
  {
    "key": "93f89fac-baa1-4095-8d26-a895077c9cbf",
    "type": "./components/Text/index.js",
  },
  {
    "key": "ba2cef2e-5ce2-43d9-8e6e-885bedfebe0e",
    "type": "./components/Button/index.js",
  },
  {
    "key": "a2c976a9-050d-4ae8-a38f-47d0fd8bfb9b",
    "type": "./components/Box/index.js",
  },
  {
    "key": "ff218c09-26b3-41e5-a8ab-ce0bb1928cfd",
    "type": "./components/Box/index.js",
  },
  {
    "key": "d7fec7c7-648a-4a4f-9575-2a89f80b3c9e",
    "type": "./components/Text/index.js",
  },
  {
    "key": "8b517842-b920-401e-851d-a895c191a1bf",
    "type": "./components/Text/index.js",
  },
  {
    "key": "0866ec28-8338-46d1-99a9-06a68b2ee0df",
    "type": "./components/Button/index.js",
  },
  {
    "key": "45872075-8891-4802-a6f2-2fcc9f5820f2",
    "type": "./components/Box/index.js",
  },
  {
    "key": "20437331-c0d7-4aa3-add0-1d872ed5fe9f",
    "type": "./components/Box/index.js",
  },
  {
    "key": "027cc40c-d276-4e55-a22e-9c2f6edaf7a5",
    "type": "./components/Text/index.js",
  },
  {
    "key": "39c5caef-71ef-4cf8-b2c8-7832d0f8b00a",
    "type": "./components/Text/index.js",
  },
  {
    "key": "fcb6c76e-46be-4db4-98bd-9cb075fba851",
    "type": "./components/SocialLinks/index.js",
  },
]
`;
